### PR TITLE
lsp.cpp: move global variables into the LSP class

### PR DESCRIPTION
**Pull request by elena-pudovkina**
Move all global variables (except for reporter) into the LSP class.
Turn processing requests into an LSP method.

### CODE REVIEW

File: lsp/lsp.cpp [hunk 1]
```diff
@@ -73,265 +73,281 @@ static const char *MethodNotFound       = "-32601";
...
-  JAST message = createResponseMessage(receivedMessage);
-  message.add("result", JSON_NULLVAL);
-  isShutDown = true;
-  sendMessage(message);
-}
+    void diagnoseFile(std::string fileUri) {
+      std::string filePath = fileUri.substr(rootUri.length() + 1, std::string::npos);
+      std::unique_ptr<Top> top(new Top);
+
+      auto fileChangesPointer = changedFiles.find(fileUri);
```
Comment (terpstra) [inline lsp/lsp.cpp:281]: I'd suggest you create a new/local LSPReporter here, and then assign the global pointer to it. In the future, we'd like to pass the reporter TO the lexer, so there is no need for it to exist for longer than this call.

File: lsp/lsp.cpp [hunk 2]
```diff
@@ -73,265 +73,281 @@ static const char *MethodNotFound       = "-32601";
...
+            callMethod(method, request);
+          }
+        }
+      }
+    }
 
+    void callMethod(std::string method, JAST request) {
+      auto functionPointer = methodToFunction.find(method);
+      if (functionPointer != methodToFunction.end()) {
+          ((*this).*(functionPointer->second))(request);
```
Comment (terpstra) [inline lsp/lsp.cpp:171]: Does `this->*(functionPointer->second)(request)` work?

**Review by terpstra — APPROVED** [review]
This looks great!

File: lsp/lsp.cpp (hunk 1 above)
Comment (elena-pudovkina) [reply lsp/lsp.cpp:281]: Done.

File: lsp/lsp.cpp (hunk 2 above)
Comment (elena-pudovkina) [reply lsp/lsp.cpp:171]: Yes, thank you

**Pull request merged**